### ISSUE TITLE
Link dependent IE tests

### ISSUE THREAD

**Issue by trainman261** [yt-dlp/yt-dlp#7482]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.06.22** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

In the InfoExtractor class (extractor/common.py) neither the methods that parse m3u8 nor the methods that extract m3u8 files have any sort of verification that what they're downloading actually _is_ an m3u8 file. If something gets downloaded that isn't an m3u8 file, the parser will still do whatever it can to parse it, leading to very confusing errors. 
In my case (see CBC Player PR, coming in a bit), this check is necessary, and the CBCPlayerIE has to check it, because sometimes the file can be a m3u8 file, sometimes an mp3 of similar. I'm thinking, though, that this isn't the only place this check might be necessary, and it might help other devs in debugging.
I'm OK with doing a PR for this, please let me know.
(Commit I'm working with: 8f05fbae2a79ce0713077ccc68b354e63216bf20)

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by pukkandan** [comment]
It's not clear to me what you are proposing - can you elaborate? How are you proposing we confirm m3u8 in an extractor independant way? Or are you suggesting a helper for extractors? Or something else?

The linked commit don't appear to have anything to do with m3u8 parsing.

**Comment by trainman261** [comment]
First off, sorry about the confusing title, had too many things in my mind at once, and I put in a title from a different issue. Stupid mistake...

As to your question: My idea is that `_parse_m3u8_formats_and_subtitles` and similar methods check to ensure that it's a valid m3u8 file. According to [https://developer.apple.com/documentation/http-live-streaming/video-on-demand-playlist-construction](this document) an m3u8 must begin with the `EXTM3U` tag. Checking that (with a seperate method in the InfoExtractor class) seems like the most senseful measure in my eyes, but I'm open to other ideas. If the tag isn't at the beginning of the file, I'm figuring the `_parse_m3u8_formats_and_subtitles` (or similar) method should return an error if it's not a valid m3u8 file and `Fatal=true`. Or does that run the risk of breaking too much stuff? (Or test and see?)

As to the commit: that was in regard to checklist item 3 - that's the commit from the master branch I was working with - only a couple of days old.

**Comment by bashonly** [comment]
> because sometimes the file can be a m3u8 file, sometimes an mp3 of similar. 

Expanding from my PR comment here: https://github.com/yt-dlp/yt-dlp/pull/7484#discussion_r1248923831

if the extractor has downloaded the non-m3u8 file (mp3/mp4), then the damage has already been done IMO. It should be the extractor's burden to check that the URL it's requesting is actually an m3u8 (e.g. via `HEADRequest`) before downloading it

**Comment by pukkandan** [comment]
Agreed. But ig adding a warning by checking `EXTM3U` tag couldn't hurt and might actually help debugging.

**Comment by dirkf** [comment]
Maybe a new Exception class that could be raised by `_parse_m3u8_formats_and_subtitles()` if it finds that the `m3u8_doc` isn't one would be better than a new method to test that?

**Comment by trainman261** [comment]
The idea with the new method was because there are multiple methods that parse m3u8 files - AFAIK `_parse_m3u8_formats_and_subtitles()` and `_parse_m3u8_formats(), maybe more. By implementing this as a new method we could just call that method. Unless that's too little to create a new method?

**Comment by bashonly** [comment]
Closed by e99c0b838a9c5feb40c0dcd291bd7b8620b8d36d
